### PR TITLE
Fax update for campaign v3

### PR DESCRIPTION
**Pull request by mcfatelin**
FAX updated for the campaign. The changes are

1. s2_area_mean_top_fraction changed to the one compatible with the LCE map used. See [#491](https://github.com/XENON1T/pax/issues/491)
2. Store the truth information based on the hits on PMTs. Previously it was based on the photons after detection but before allocating the PMT hits. 
3. Add the top_fraction information into truth output.
4. Enable the PMT to generate after pulse > 1PE. The amplitude of AP comes from study by Julien.

### CODE REVIEW

File: pax/simulation.py [hunk 1]
```diff
@@ -314,7 +338,7 @@ def make_pax_event(self):
                                                         n_rvs=n_afterpulses))
 
             # If none of the setting specified, do not make afterpulses
-
+            ap_gains = [ x*y for x, y in zip(ap_gains, ap_amplifications) ]
```
Comment (tunnell) [inline pax/simulation.py]: __[flake8]__
* [E201] whitespace after '['
* [E202] whitespace before ']'

File: pax/plugins/io/WaveformSimulator.py [hunk 2]
```diff
@@ -118,9 +118,25 @@ def s2(self, electrons, g4_id=-1, t=0., x=0., y=0., z=0.):
             x = r * np.cos(phi)
             y = r * np.sin(phi)
 
-        return self.simulator.queue_signal(photon_times, x, y,
+        arriving_photon_times_top,  arriving_photon_times_bottom = self.simulator.queue_signal(photon_times, x, y,
                                            # Generate S2 hitpattern "at the anode": cue for simulator to use S2 LCE map
                                            z=-self.config['gate_to_anode_distance'])
+        peak_top_fraction = float(len(arriving_photon_times_top)) / float(
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [F841] local variable 'peak_top_fraction' is assigned to but never used

File: pax/plugins/io/WaveformSimulator.py [hunk 3]
```diff
@@ -118,9 +118,25 @@ def s2(self, electrons, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+            len(arriving_photon_times_bottom) + len(arriving_photon_times_top) )
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E202] whitespace before ')'

File: pax/plugins/io/WaveformSimulator.py [hunk 4]
```diff
@@ -118,9 +118,25 @@ def s2(self, electrons, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+        self.store_true_peak('s2',
+                                          g4_id,
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: pax/plugins/io/WaveformSimulator.py [hunk 5]
```diff
@@ -118,9 +118,25 @@ def s2(self, electrons, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                          t,
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: pax/plugins/io/WaveformSimulator.py [hunk 6]
```diff
@@ -118,9 +118,25 @@ def s2(self, electrons, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                          x,
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: pax/plugins/io/WaveformSimulator.py [hunk 7]
```diff
@@ -118,9 +118,25 @@ def s2(self, electrons, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                          y,
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: pax/plugins/io/WaveformSimulator.py [hunk 8]
```diff
@@ -118,9 +118,25 @@ def s2(self, electrons, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                          z,
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: pax/plugins/io/WaveformSimulator.py [hunk 9]
```diff
@@ -118,9 +118,25 @@ def s2(self, electrons, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                          np.concatenate((
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: pax/plugins/io/WaveformSimulator.py [hunk 10]
```diff
@@ -118,9 +118,25 @@ def s2(self, electrons, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                                                   arriving_photon_times_top,
+                                                                   arriving_photon_times_bottom
+                                                                   )
+                                                                  ),
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E124] closing bracket does not match visual indentation

File: pax/plugins/io/WaveformSimulator.py [hunk 11]
```diff
@@ -118,9 +118,25 @@ def s2(self, electrons, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                          electron_times,
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: pax/plugins/io/WaveformSimulator.py [hunk 12]
```diff
@@ -118,9 +118,25 @@ def s2(self, electrons, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                          )
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E124] closing bracket does not match visual indentation

File: pax/plugins/io/WaveformSimulator.py [hunk 13]
```diff
@@ -133,10 +149,30 @@ def s1(self, photons, recoil_type, g4_id=-1, t=0., x=0., y=0., z=0.):
         photon_times = self.simulator.s1_photons(photons, recoil_type, x, y, z, t)
         if not len(photon_times):
             return None
-        self.store_true_peak('s1', g4_id, t, x, y, z, photon_times)
-        return self.simulator.queue_signal(photon_times, x=x, y=y, z=z)
-
-    def s2_after_pulses(self):
+        arriving_photon_times_top, arriving_photon_times_bottom =  self.simulator.queue_signal(
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E222] multiple spaces after operator

File: pax/plugins/io/WaveformSimulator.py [hunk 14]
```diff
@@ -133,10 +149,30 @@ def s1(self, photons, recoil_type, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                                                                               photon_times,
+                                                                                               x=x,
+                                                                                               y=y,
+                                                                                               z=z
+                                                                                              )
+        peak_top_fraction = float(len(arriving_photon_times_top)) / float(
+            len(arriving_photon_times_bottom) + len(arriving_photon_times_top) )
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E202] whitespace before ')'

File: pax/plugins/io/WaveformSimulator.py [hunk 15]
```diff
@@ -133,10 +149,30 @@ def s1(self, photons, recoil_type, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+        self.store_true_peak('s1',
+                                         g4_id,
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: pax/plugins/io/WaveformSimulator.py [hunk 16]
```diff
@@ -133,10 +149,30 @@ def s1(self, photons, recoil_type, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                         t,
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: pax/plugins/io/WaveformSimulator.py [hunk 17]
```diff
@@ -133,10 +149,30 @@ def s1(self, photons, recoil_type, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                         x,
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: pax/plugins/io/WaveformSimulator.py [hunk 18]
```diff
@@ -133,10 +149,30 @@ def s1(self, photons, recoil_type, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                         y,
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: pax/plugins/io/WaveformSimulator.py [hunk 19]
```diff
@@ -133,10 +149,30 @@ def s1(self, photons, recoil_type, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                         z,
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: pax/plugins/io/WaveformSimulator.py [hunk 20]
```diff
@@ -133,10 +149,30 @@ def s1(self, photons, recoil_type, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                         np.concatenate((
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: pax/plugins/io/WaveformSimulator.py [hunk 21]
```diff
@@ -133,10 +149,30 @@ def s1(self, photons, recoil_type, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                                                   arriving_photon_times_top,
+                                                                   arriving_photon_times_bottom
+                                                                   )
+                                                                  ),
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E124] closing bracket does not match visual indentation

File: pax/plugins/io/WaveformSimulator.py [hunk 22]
```diff
@@ -133,10 +149,30 @@ def s1(self, photons, recoil_type, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                         peak_top_fraction = peak_top_fraction,
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent
* [E251] unexpected spaces around keyword / parameter equals

File: pax/plugins/io/WaveformSimulator.py [hunk 23]
```diff
@@ -133,10 +149,30 @@ def s1(self, photons, recoil_type, g4_id=-1, t=0., x=0., y=0., z=0.):
...
+                                         )
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E124] closing bracket does not match visual indentation

File: pax/plugins/io/WaveformSimulator.py [hunk 24]
```diff
@@ -191,12 +227,27 @@ def s2_after_pulses(self):
                 Y[electron_id]
                 )
             # queue the photons caused by the s2 after pulses
-            self.simulator.queue_signal(
+            arriving_photon_times_top, arriving_photon_times_bottom =  self.simulator.queue_signal(
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E222] multiple spaces after operator

File: pax/plugins/io/WaveformSimulator.py [hunk 25]
```diff
@@ -191,12 +227,27 @@ def s2_after_pulses(self):
...
                 s2_ap_photon_times,
                 X[electron_id],
                 Y[electron_id],
                 -self.config['gate_to_anode_distance']
                 )
+            peak_top_fraction = float(len(arriving_photon_times_top)) / float(
+                len(arriving_photon_times_bottom) + len(arriving_photon_times_top) )
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E202] whitespace before ')'

File: pax/plugins/io/WaveformSimulator.py [hunk 26]
```diff
@@ -191,12 +227,27 @@ def s2_after_pulses(self):
...
+            self.store_true_peak('photoionization_afterpulse',
+                                          g4_id,
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: pax/plugins/io/WaveformSimulator.py [hunk 27]
```diff
@@ -191,12 +227,27 @@ def s2_after_pulses(self):
...
+                                          t = s2_ap_electron_time,
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent
* [E251] unexpected spaces around keyword / parameter equals

File: pax/plugins/io/WaveformSimulator.py [hunk 28]
```diff
@@ -191,12 +227,27 @@ def s2_after_pulses(self):
...
+                                          x = X[electron_id],
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent
* [E251] unexpected spaces around keyword / parameter equals

File: pax/plugins/io/WaveformSimulator.py [hunk 29]
```diff
@@ -191,12 +227,27 @@ def s2_after_pulses(self):
...
+                                          y = Y[electron_id],
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent
* [E251] unexpected spaces around keyword / parameter equals

File: pax/plugins/io/WaveformSimulator.py [hunk 30]
```diff
@@ -191,12 +227,27 @@ def s2_after_pulses(self):
...
+                                          z = -self.config['gate_to_anode_distance'],
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent
* [E251] unexpected spaces around keyword / parameter equals

File: pax/plugins/io/WaveformSimulator.py [hunk 31]
```diff
@@ -191,12 +227,27 @@ def s2_after_pulses(self):
...
+                                          photon_times = np.concatenate((
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent
* [E251] unexpected spaces around keyword / parameter equals

File: pax/plugins/io/WaveformSimulator.py [hunk 32]
```diff
@@ -191,12 +227,27 @@ def s2_after_pulses(self):
...
+                                                                   arriving_photon_times_top,
+                                                                   arriving_photon_times_bottom
+                                                                   )
+                                                                  ),
+                                          peak_top_fraction = peak_top_fraction,
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent
* [E251] unexpected spaces around keyword / parameter equals

File: pax/plugins/io/WaveformSimulator.py [hunk 33]
```diff
@@ -191,12 +227,27 @@ def s2_after_pulses(self):
...
+                                          )
```
Comment (tunnell) [inline pax/plugins/io/WaveformSimulator.py]: __[flake8]__
* [E124] closing bracket does not match visual indentation

**Comment by pdeperio** [comment]
Should partially fix https://github.com/XENON1T/mc/issues/20

**Comment by pdeperio** [comment]
Discussed Jelle's two points in person:
  * Function is aggregating the info into a few variables just like before, not actually storing each photon
  * With the option to avoid dead PMTs during per-PMT distribution, current truth storage is consistent before and after considering dead PMTs (since total photons already scaled from @berget2's data-derived LCE map including dead PMTs)

Will need to reconsider all of this when updating to MC-only maps in the future (for both S1 and S2).

The [processing scripts](https://github.com/XENON1T/processing/tree/master/montecarlo) must be synchronized for new truth info (@jhowl01 in arrays output, @mcfatelin in flat output) and removing ```MEAN_TOP_FRACTION``` hack.

**Pull request merged**